diracx-db/src/diracx/db/sql/job/db.py

```diff
@@ -189,243 +206,187 @@ async def getJobJDL(self, job_id: int, original: bool = False) -> str:
+                    "SubmissionTime": datetime.now(tz=timezone.utc),
+                    "Owner": job.owner,
+                    "OwnerGroup": job.owner_group,
+                    "VO": job.vo,
+                }
+
+                jobManifest_.setOption("JobID", job_id)
+
+                # 2.- Check JDL and Prepare DIRAC JDL
+                jobJDL = jobManifest_.dumpAsJDL()
+
+                # Replace the JobID placeholder if any
+                if jobJDL.find("%j") != -1:
+                    jobJDL = jobJDL.replace("%j", str(job_id))
+
+                class_ad_job = ClassAd(jobJDL)
+
+                class_ad_req = ClassAd("[]")
+                if not class_ad_job.isOK():
+                    # Rollback the entire transaction
+                    raise ValueError(f"Error in JDL syntax for job JDL: {original_jdl}")
+                # TODO: check if that is actually true
+                if class_ad_job.lookupAttribute("Parameters"):
+                    raise NotImplementedError("Parameters in the JDL are not supported")
+
+                # TODO is this even needed?
+                class_ad_job.insertAttributeInt("JobID", job_id)
+
+                await self.checkAndPrepareJob(
+                    job_id,
+                    class_ad_job,
+                    class_ad_req,
+                    job.owner,
+                    job.owner_group,
+                    job_attrs,
+                    job.vo,
+                )
+                jobJDL = createJDLWithInitialStatus(
+                    class_ad_job,
+                    class_ad_req,
+                    self.jdl2DBParameters,
+                    job_attrs,
+                    job.initial_status,
+                    job.initial_minor_status,
+                    modern=True,
+                )
+                # assert "JobType" in job_attrs, job_attrs
+                job_ids.append(job_id)
+                jobs_to_insert.append(job_attrs)
+                jdls_to_update.append(
+                    {
+                        "b_JobID": job_id,
+                        "JDL": compressJDL(jobJDL),
+                    }
+                )
 
-        if reschedule_counter > self.maxRescheduling:
-            logging.warn(f"Job {job_id}: Maximum number of reschedulings is reached.")
-            self.setJobAttributes(
-                job_id,
-                {
-                    "Status": JobStatus.FAILED,
-                    "MinorStatus": JobMinorStatus.MAX_RESCHEDULING,
-                },
-            )
-            raise ValueError(
-                f"Maximum number of reschedulings is reached: {self.maxRescheduling}"
+                if class_ad_job.lookupAttribute("InputData"):
+                    inputData = class_ad_job.getListFromExpression("InputData")
+                    inputdata_to_insert += [
+                        {"JobID": job_id, "LFN": lfn} for lfn in inputData if lfn
+                    ]
```

Comment:
I am wondering whether it makes sense to have all this logic in the data access layer rather than in the router (at least part of it).

I haven't checked carefully, but it looks like the part dealing with `JobManifest` and `ClassAd` have not much to do with the DB, am I wrong?

Reply:
I think that sounds very sensible to me - are you happy for me to move all of this into the router, or perhaps into a utility module in the routers?

Reply:
IIUC:

- This bloc of code could move to the router. Then you would probably call a DB method to insert the original JDLs https://github.com/DIRACGrid/diracx/blob/1fb0f3aca4363c53eb1184880710f68baca40fba/diracx-db/src/diracx/db/sql/job/db.py#L237-L252
- Then the following bloc of code could also go into the router, and may become a utility function such as `check_and_prepare_jdl` (you would remove the `checkAndPrepareJob` function from the `db` class and would add its content into the new `check_and_prepare_jdl`): https://github.com/DIRACGrid/diracx/blob/1fb0f3aca4363c53eb1184880710f68baca40fba/diracx-db/src/diracx/db/sql/job/db.py#L270-L340
- The remaining lines are responsible for inserting the JDL, the job and its inputs into the DB, and should become the content of `insert_bulk` I would say.

What do you think?

Reply:
Done, but JobDB tests relied on most or all of this logic being inside job DB... can't import the utility fn from routers because of (presumably) a cyclic dependency. Even more confusing as everything runs just fine locally for me. It is very tempting to `pytest.skip`... but it could be something we could resolve by tweaking the CI environment too. what do now?

Reply:
We agreed to keep these as utility functions inside `diracx-db` in case agents will need to call them (i.e. not via the API/services)